winit-wayland/src/seat/text_input/mod.rs

```diff
@@ -115,12 +115,42 @@ impl Dispatch<ZwpTextInputV3, TextInputData, WinitState> for TextInputState {
                 text_input_data.pending_preedit = None;
                 text_input_data.pending_commit = text;
             },
+            TextInputEvent::DeleteSurroundingText { before_length, after_length } => {
+                text_input_data.pending_delete = Some(DeleteSurroundingText {
+                    before: before_length as usize,
+                    after: after_length as usize,
+                });
+            },
             TextInputEvent::Done { .. } => {
                 let window_id = match text_input_data.surface.as_ref() {
                     Some(surface) => crate::make_wid(surface),
                     None => return,
                 };
 
+                // The events are sent to the user separately, so
+                // CAUTION: events must always arrive in the order compatible with the application
+                // order specified by the text-input-v3 protocol:
+                //
+                // As of version 1:
+                // 1. Replace existing preedit string with the cursor.
+                // 2. Delete requested surrounding text.
+                // 3. Insert commit string with the cursor at its end.
+                // 4. Calculate surrounding text to send.
+                // 5. Insert new preedit text in cursor position.
+                // 6. Place cursor inside preedit text.
```

Comment:
Maybe it would be good to have this in top-level IME docs? I'd assume that it's the same on other systems, since I don't see how it could be done differently.

Reply:
Added a note about applying in the order they arrived.